### PR TITLE
chore: Analytic events write back + refactor

### PR DESCRIPTION
**Pull request by rephus**
Closes:   #11961

### Description:

Screenshot from 2024-10-18 16-54-07
Screenshot from 2024-10-18 16-55-26
Screenshot from 2024-10-18 16-56-40

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/backend/src/analytics/LightdashAnalytics.ts [hunk 1]
```diff
@@ -1057,7 +1057,11 @@ export type SemanticLayerView = BaseTrack & {
     };
 };
 
-export type BaseVirtualViewEvent = BaseTrack & {
```
Comment (rephus) [inline packages/backend/src/analytics/LightdashAnalytics.ts:1060]: I simplified these types a bit

File: packages/backend/src/controllers/githubController.ts [hunk 2]
```diff
@@ -48,25 +39,17 @@ export class GithubInstallController extends BaseController {
...
     ): Promise<void> {
-        const returnToUrl = new URL(
-            '/generalSettings/integrations',
-            lightdashConfig.siteUrl,
-        );
-        const randomID = nanoid().replace('_', ''); // we use _ as separator, don't allow this character on the nanoid
-        const subdomain = lightdashConfig.github.redirectDomain;
-        const state = `${subdomain}_${randomID}`;
-        const githubAppName = lightdashConfig.github.appName;
+        const context = await this.services
```
Comment (rephus) [inline packages/backend/src/controllers/githubController.ts:42]: moved logic to service

File: packages/backend/src/services/GithubAppService/GithubAppService.ts [hunk 3]
```diff
@@ -1,30 +1,159 @@
...
+        return {
+            installUrl: `https://github.com/apps/${githubAppName}/installations/new?state=${state}`,
+            returnToUrl: returnToUrl.href,
+            state,
+            githubAppName,
+            inviteCode: user.userUuid,
+        };
+    }
+
+    async installCallback(
```
Comment (rephus) [inline packages/backend/src/services/GithubAppService/GithubAppService.ts:75]: copied from the controller

File: packages/backend/src/services/GithubAppService/GithubAppService.ts [hunk 4]
```diff
@@ -1,30 +1,159 @@
...
+                throw new Error('Invalid installation id');
+
+            await this.upsertInstallation(
+                userUuid,
+                installation_id,
+                token,
+                refreshToken,
+            );
+
+            this.analytics.track({
```
Comment (rephus) [inline packages/backend/src/services/GithubAppService/GithubAppService.ts:136]: this is the only addition to the logic, tracking the event

File: packages/backend/src/controllers/githubController.ts (hunk 2 above)
Comment (joaoviana) [reply packages/backend/src/controllers/githubController.ts:42]: 🙌 super clean

File: packages/backend/src/services/GitIntegrationService/GitIntegrationService.ts [hunk 5]
```diff
@@ -662,6 +668,16 @@ Triggered by user ${user.firstName} ${user.lastName} (${user.email})
             base: branch,
         });
 
+        this.analytics.track({
```
Comment (joaoviana) [inline packages/backend/src/services/GitIntegrationService/GitIntegrationService.ts:671]: Screenshot 2024-10-21 at 09 19 48

**Review by joaoviana — APPROVED** [review]
started

Screenshot 2024-10-21 at 09 21 06

completed

Screenshot 2024-10-21 at 09 21 40

**Pull request merged**